[Offload] Have olMemFree accept a platform as a param

RossBrunton opened the pull request:
In a future change, most of the allocation tracking will be removed from
liboffload itself and be delegated to the plugins. Therefore, we will
need to know which plugin is in charge of the allocation.

RossBrunton committed:
[Offload] Have olMemFree accept a platform as a param

In a future change, most of the allocation tracking will be removed from
liboffload itself and be delegated to the plugins. Therefore, we will
need to know which plugin is in charge of the allocation.

leandrolcampos commented on offload/unittests/Conformance/include/mathtest/DeviceContext.hpp:
```diff
@@ -57,13 +57,13 @@ class DeviceContext {
 
   template <typename T>
-  ManagedBuffer<T> createManagedBuffer(std::size_t Size) const noexcept {
+  ManagedBuffer<T> createManagedBuffer(std::size_t Size) noexcept {
```
Hi @RossBrunton,
I'll take a look until the end of the day.

leandrolcampos replied:
@RossBrunton, could you take a look at this #157773? Merging this patch will ensure the conformance tests run correctly.

I also prepared a [Gist](https://gist.github.com/leandrolcampos/091bb5ac6c4521de895888e016a251cb) containing a step-by-step guide to setting up an environment to build and run the conformance tests on Windows Subsystem for Linux (WSL 2) with NVIDIA GPUs. I think it can be easily adapted to different OS and/or platform configurations.

Finally, thanks for your care with the conformance tests!

leandrolcampos committed:
[Offload][Conformance] Update olMemFree calls in conformance tests (#157773)

This PR is a follow-up to the change introduced in #157478, which added
a `platform` parameter to the `olMemFree` function.

RossBrunton commented:
@jhuber6 @callumfare Can you give this another look over when you get the chance?

pbalcer reviewed (commented):
This looks good to me, but one problem I've noticed in the interface that a single plugin can only expose a single platform. This is not true in general. For example, in level-zero, there can be multiple separate drivers in the OS, handling different devices. These need to be separate platforms.

jhuber6 reviewed (commented):
I would still really prefer if we didn't need to change the free interface. As far as I understand, the general problem is that when we start working with heterogeneous devices, we could have two devices that return the same address and thus confound our attempts to map pointers to devices.

The only time this can occur is for allocations on the device itself, managed allocations will go through the shared Linux kernel. If I understand correctly, after allocation we will put something into some kind of map. If we already have a map, then we should be able to detect if there is a conflict. If there is a conflict, we should just allocate the memory again, try adding it again, then free the old pointer. It's slightly inefficient, but this is a one in a million edge case, it does not need to be fast. Does this make sense to you?

RossBrunton commented:
@jhuber6 @pbalcer To summarize the discussion in the meeting:

We need the ability to, given a pointer to somewhere in memory, figure out which platform allocated it so that we can dispatch to the appropriate platform for olMemFree/olGetMemInfo and friends.

There are (at least) two possible approaches:
1. We store a list of allocations in liboffload, along with their size, device and maybe some other information. This is an ordered list, and we can do a binary search to find which memory allocation a pointer is in. When we allocate, we check to see if the new allocation overlaps another one. If it does, we make a new allocation and throw away the old one. This ensures that every address is associated with at most one device/platform.
2. We require the liboffload user to track the device/platform, and pass that as a parameter to MemFree and GetMemInfo. This means that liboffload doesn't need to keep its own list of allocations (they will de-facto be managed by the user, which may have been tracking it anyway). GetInfo queries themselves can be done in a platform-specific way which will hopefully be more performant.

Performance wise, option 1 requires a binary search on every allocation, free and info query. Hopefully collisions are rare enough that they don't have an impact on performance. Option 2's performance will depend on how the user keeps track of platforms, and could be faster if they can access the platform handle in constant time.

Ergonomics wise, option 1 is much simpler, since the user doesn't have to track metadata about their allocations.

I don't remember if this got mentioned in the meeting, but I think there was an idea for having the runtime query each platform in turn to see if they owned the pointer. I don't think this will work because unused platforms might be in some uninitialized or sleeping state. So sending information queries might provoke the driver into going and waking up a GPU or something. And that's assuming that the platform driver is not bugged in such a way that it "claims" memory it hasn't allocated.
